app/lib/company_shareholders_updator.rb

```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class CompanyShareholdersUpdator < ApplicationService
+  def initialize
+    @gpw_client = GPW::Client.new
+  end
+
+  def call(isin)
+    company = Company.find_by(isin: isin)
+    company_shareholders = @gpw_client.fetch_company_shareholders(company.stock, isin)
+    update_company_shareholders(company_shareholders, isin)
+  end
+
+  private
+
+  def update_company_shareholders(company_shareholders, isin)
+    company = Company.find_by(isin: isin)
+    company_shareholders.each do |shareholder_attrs|
+      shareholder = company.shareholders.find_or_create_by!(name: shareholder_attrs.fetch(:name))
+      shares_count = shareholder_attrs.fetch(:shares_count)
+      update_shares_count(company, shareholder, shares_count)
```

Comment:
Probably yes, but I think, not shares, I should **upsert** shareholders, because shares are already created together with shareholders in line 18, so in line 16 should be:
`def upsert_company_shareholders(company_shareholders, company)`

Reply:
ok, I would probably write it in a way that shares are updated in a single query (as I undesrstant that right now you create some shares in line 18 and later update count on them)

Reply:
but I would need to think a bit more and probably understand better what you want to do to provide any specyfic example... so maybe you can just keep your code as it is for now if it works "good enough" :)

Reply:
I think we need to get short took about, but maybe on the weekend.. :)